### PR TITLE
AP_RangFinder: various maxbotix serial sonar

### LINKED ISSUE #7885: Maxbotix Serial in mm
#### Issue details :
Some Maxbotix Serial (MB1043 HRLV-MaxSonar-EZ4) are in mm not inch.
[https://www.maxbotix.com/Ultrasonic_Sensors/MB1043.htm](https://www.maxbotix.com/Ultrasonic_Sensors/MB1043.htm)
Why not using parameter "RNGFND_RMETRIC" to enable or not code "reading_cm = 2.54f * sum / count;"
Perhaps another parameter can be use to adjust altitude offset (if sonar not at bottom location).

#### Version :
latests commit of [ardupilot/libraries/AP_RangeFinder/AP_RangeFinder_MaxsonarSerialLV.cpp](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_RangeFinder/AP_RangeFinder_MaxsonarSerialLV.cpp)
commit [link](https://github.com/ArduPilot/ardupilot/commit/146143ff2ed04808fffeac79371402cd8b954672)

#### Platform :
[  ] All
[  ] AntennaTracker
[X] Copter
[  ] Plane
[  ] Rover
[  ] Submarine
but this is a lib, perhaps others can benefit.

#### Airframe type
Doesn't matter

#### Hardware type
Navio2 and others.

#### Logs
No logs

Regards.

### LINKED ISSUE #16749: Support maxbotix ezl4 lidar
I am using an omnibus f4 pro that has very limited analog pins, however it has uarts to spare. If there was support for connecting a serial lidar sensor such as the maxbotix ezl4, then that would greatly increase the functionality of this fc

this would be good for quads or planes

### PR DESCRIPTION
**Pull request by chobitsfan**
fix #7885 and fix #16749

Maxbotix serial sonar may report reading in inch, cm or mm. Current implementation assume it gives the metrics in inches. 
This PR add 2 additional rangfinder type: 35 for MaxbotixSerial cm reading and 36 for MaxbotixSerial mm reading.

P.S: some Maxbotix serial sonar (MB1240, for example) may need to set InvertRX in SERIALx_OPTIONS 

Test ok in MB1240 (cm reading)

### CODE REVIEW

File: libraries/AP_RangeFinder/AP_RangeFinder_MaxsonarSerialLV.cpp [hunk 1]
```diff
@@ -56,7 +70,7 @@ bool AP_RangeFinder_MaxsonarSerialLV::get_reading(float &reading_m)
     }
 
     // This sonar gives the metrics in inches, so we have to transform this to meters
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_MaxsonarSerialLV.cpp:72]: Comment needs fixing
\

**Review by peterbarker — COMMENTED** [review]
This looks correct, but may not be the right way to fix it.

The other way to fix it is to add a parameter sub-tree just for this rangefinder - we add one for the Wasp backend, for example.

Marking DevCall for discussion.

Note that since these are new rangefinder types - and we test every rangefinder except CAN in SITL - adding a pair of simulated rangefinders for cm/mm would be good.

File: libraries/AP_RangeFinder/AP_RangeFinder_MaxsonarSerialLV.cpp (hunk 1 above)
Comment (chobitsfan) [reply libraries/AP_RangeFinder/AP_RangeFinder_MaxsonarSerialLV.cpp]: Thank you for reminding.

**Comment by chobitsfan** [comment]
Hi @peterbarker Thank you for reviewing

> The other way to fix it is to add a parameter sub-tree just for this rangefinder - we add one for the Wasp backend, for example.

Yes, I agree. But this parameter is only used by maxbotix serial sonar.  I think maybe it is not really necessary to add a new parameter just for one specific sensor driver.

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
if we do scaling.set_default(1.0) in the maxbotix driver, then apply the scaling, users can change scaling as needed

**Comment by peterbarker** [comment]
It was suggested on the Devcall we use scaling factor parameter instead.  We didn't decide to ask for that, 'though, and we'll probably merge this after docs fixed.

File: libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp [hunk 2]
```diff
@@ -6,7 +6,7 @@ const AP_Param::GroupInfo AP_RangeFinder_Params::var_info[] = {
...
     // @DisplayName: Rangefinder type
     // @Description: What type of rangefinder device that is connected
-    // @Values: 0:None,1:Analog,2:MaxbotixI2C,3:LidarLite-I2C,5:PWM,6:BBB-PRU,7:LightWareI2C,8:LightWareSerial,9:Bebop,10:MAVLink,11:USD1_Serial,12:LeddarOne,13:MaxbotixSerial,14:TeraRangerI2C,15:LidarLiteV3-I2C,16:VL53L0X or VL53L1X,17:NMEA,18:WASP-LRF,19:BenewakeTF02,20:Benewake-Serial,21:LidarLightV3HP,22:PWM,23:BlueRoboticsPing,24:DroneCAN,25:BenewakeTFminiPlus-I2C,26:LanbaoPSK-CM8JL65-CC5,27:BenewakeTF03,28:VL53L1X-ShortRange,29:LeddarVu8-Serial,30:HC-SR04,31:GYUS42v2,32:MSP,33:USD1_CAN,34:Benewake_CAN,100:SITL
+    // @Values: 0:None,1:Analog,2:MaxbotixI2C,3:LidarLite-I2C,5:PWM,6:BBB-PRU,7:LightWareI2C,8:LightWareSerial,9:Bebop,10:MAVLink,11:USD1_Serial,12:LeddarOne,13:MaxbotixSerial,14:TeraRangerI2C,15:LidarLiteV3-I2C,16:VL53L0X or VL53L1X,17:NMEA,18:WASP-LRF,19:BenewakeTF02,20:Benewake-Serial,21:LidarLightV3HP,22:PWM,23:BlueRoboticsPing,24:DroneCAN,25:BenewakeTFminiPlus-I2C,26:LanbaoPSK-CM8JL65-CC5,27:BenewakeTF03,28:VL53L1X-ShortRange,29:LeddarVu8-Serial,30:HC-SR04,31:GYUS42v2,32:MSP,33:USD1_CAN,34:Benewake_CAN,35:MaxbotixSerialCM,36:MaxbotixSerialMM,100:SITL
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp:9]: In the parameter descriptions can you group the 3 items, "13:MaxbotixSerial", 35 and 36 together?  Also let's clarify the units of the existing option (13) by changing its name to include the units.

Personally I agree with @peterbarker that reusing the SCALING parameter would be a good alternative but I think we're ok with this change as well.

**Comment by chobitsfan** [comment]
> It was suggested on the Devcall we use scaling factor parameter instead. We didn't decide to ask for that, 'though, and we'll probably merge this after docs fixed.

Hi @tridge , @rmackay9 and @peterbarker Thank you for suggestion. I agree with you. It is a better to use scaling factor parameter instead. I will change to this approach.

**Comment by chobitsfan** [comment]
replaced by #19778

**Pull request closed without merging**